### PR TITLE
FIX: Remove magic time ranges

### PR DESCRIPTION
**Pull request by Drenmi**
### History

In #15474 we introduced dedicated support for date ranges. As part of that change we added a fallback of "magic" date ranges, which treats dates in any paragraph with _exactly_ two dates as a range. There were discussions about migrating all such paragraphs to use the new date range element, but it was ultimately decided against.

### What's the problem?

1️⃣ A minor [bug](https://meta.discourse.org/t/timezone-bug/248674) was reported relating to the "magic" date range. If a Tippy was already open when expanding a date, the date would be considered a single date, but if no Tippy was open, it would be considered a range. So expanding the same date could have different results based on the initial conditions.

(The reason this happened is the fallback relies on the number of elements in the paragraph being exactly 2, but the Tippy element was also added as a child of the same paragraph, resulting in 3 children.)

2️⃣ While working on this, I additionally found that using two date ranges in the same paragraph would result in all dates in the paragraph being treated as single dates, rather than being grouped together in ranges.

### How does this fix it?

Firstly, we remove the fallback to "magic" date ranges, as decided. This fixes the 1️⃣  problem.

Secondly, when looking at an element and it's parent paragraph, we now select all date range children and partition them two-by-two. This fixes the 2️⃣ problem.

### In summary

- Single dates are now always rendered as single dates.
- Multiple date ranges in the same paragraph now works correctly.

### In action

discourse-date-ranges

### CODE REVIEW

File: plugins/discourse-local-dates/assets/javascripts/initializers/discourse-local-dates.js [hunk 1]
```diff
@@ -124,18 +124,24 @@ function _rangeElements(element) {
...
+  if (element.dataset.range) {
+    return _partitionedRanges(element).find((pair) => pair.includes(element));
   }
 
-  return Array.from(element.parentElement.children).filter(
-    (span) => span.dataset.date
+  return [element];
+}
+
+function _partitionedRanges(element) {
```
Comment (Drenmi) [inline plugins/discourse-local-dates/assets/javascripts/initializers/discourse-local-dates.js:134]: This function assumes date-range elements always come in pairs of two. I decided not to be defensive about it, as there is nothing we could do to recover at this point if something is wrong upstream.

**Review by martin-brennan — APPROVED** [review]
The fix looks good, I am not sure if we can add to the existing qunit acceptance test for this? Even if it was just for the multiple ranges in a paragraph issue.

**Comment by Drenmi** [comment]
@martin-brennan Since all the QUnit setup was for testing the composer, and I had to do the setup anyway, I opted for Rails system specs instead. 🙇

File: plugins/discourse-local-dates/spec/system/local_dates_spec.rb [hunk 2]
```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+
+describe "Local dates", type: :system, js: true do
+  fab!(:topic) { Fabricate(:topic) }
+    fab!(:user) { Fabricate(:user) }
+    fab!(:post) { Fabricate(:post_with_local_dates, topic: topic) }
+
+    it "renders local dates and date ranges correctly" do
```
Comment (martin-brennan) [inline plugins/discourse-local-dates/spec/system/local_dates_spec.rb:8]: For this to work on the CI, I think you will need to set `user.user_option.timezone` to `"Asia/Singapore"`

File: plugins/discourse-local-dates/spec/system/local_dates_spec.rb [hunk 3]
```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+
+describe "Local dates", type: :system, js: true do
+  fab!(:topic) { Fabricate(:topic) }
```
Comment (martin-brennan) [inline plugins/discourse-local-dates/spec/system/local_dates_spec.rb:4]: Weird indenting here. With new stuff, generally running syntax_tree on the code is fine (DM me internally for details if needed)

File: plugins/discourse-local-dates/spec/system/local_dates_spec.rb [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    fab!(:user) { Fabricate(:user) }
+    fab!(:post) { Fabricate(:post_with_local_dates, topic: topic) }
+
+    it "renders local dates and date ranges correctly" do
+      sign_in user
+
+      visit "/t/#{topic.id}"
+      topic_page = PageObjects::Pages::Topic.new
```
Comment (martin-brennan) [inline plugins/discourse-local-dates/spec/system/local_dates_spec.rb:12]: This has a `visit_topic` helper already

Comment (martin-brennan) [reply plugins/discourse-local-dates/spec/system/local_dates_spec.rb:12]: Also generally we have been storing these at the top of the spec in `let` blocks

File: spec/fabricators/post_fabricator.rb [hunk 5]
```diff
@@ -60,6 +60,20 @@
 HTML
 end
 
+Fabricator(:post_with_local_dates, from: :post) do
+  raw <<~RAW
```
Comment (martin-brennan) [inline spec/fabricators/post_fabricator.rb:64]: I do worry this Fabricator is overly specific and not reusable -- for instance I wonder if it was used twice in the same spec, maybe it would error because the content is the same? Also this is a plugin fabricator so it should not be defined in core (for chat we have put the plugin ones in plugins/chat/spec/fabricators/chat_fabricator.rb). Finally, we shouldn't directly apply the `cooked` inside the fabricator since it can get out of date based on markdown rules. Better to just do `post.rebake!` inside the spec, or even better use the `create_post` helper which does that for you https://github.com/discourse/discourse/blob/08eeaa78d055b2e7868e19fa45a39034b4f857fc/spec/support/helpers.rb#L47-L52

File: plugins/discourse-local-dates/spec/system/local_dates_spec.rb [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+      post_dates = topic_page.find_all("span[data-date]")
```
Comment (martin-brennan) [inline plugins/discourse-local-dates/spec/system/local_dates_spec.rb:14]: Before doing this you may want to make sure you are on the right page by checking the page has the post content e.g. `expect(page).to have_content("Date range post test")` or something similar, since capybara will wait for that but it will not wait for `find_all`

File: spec/support/system_helpers.rb [hunk 7]
```diff
@@ -40,4 +40,16 @@ def resize_window(width: nil, height: nil)
   ensure
     page.driver.browser.manage.window.resize_to(original_size.width, original_size.height)
   end
+
+  def using_browser_timezone(timezone, &example)
```
Comment (martin-brennan) [inline spec/support/system_helpers.rb:44]: Wow this is cool, I didn't know about `using_session` and I also didn't know we had the TZ in env!

**Review by martin-brennan — APPROVED** [review]
Nice one, thanks for the changes :pray:
